### PR TITLE
Implement json serialize for l10n string

### PR DESCRIPTION
**Pull request by BernhardPosselt**
Otherwise adding a translated string to a JsonResponse will return an empty value

@th3fallen @DeepDiver1975 @MorrisJobke @PVince81

### CODE REVIEW

File: lib/private/l10n/string.php [hunk 1]
```diff
@@ -74,4 +76,10 @@ public function __toString() {
 		$text = str_replace('%n', $this->count, $text);
 		return vsprintf($text, $this->parameters);
 	}
+
+
+	public function jsonSerialize() {
+		return $this->__toString();
```
Comment (th3fallen) [inline lib/private/l10n/string.php]: could just `return (string)$this;` but this works too lol

**Comment by th3fallen** [comment]
:+1: Works perfectly here

File: lib/private/l10n/string.php [hunk 2]
```diff
@@ -74,4 +76,10 @@ public function __toString() {
 		$text = str_replace('%n', $this->count, $text);
 		return vsprintf($text, $this->parameters);
 	}
+
+
```
Comment (th3fallen) [inline lib/private/l10n/string.php]: Actually one change, add the correct docblock for this method 
```
	/**
	 * (PHP 5 >= 5.4.0)<br/>
	 * Specify data which should be serialized to JSON
	 *
	 * @link http://php.net/manual/en/jsonserializable.jsonserialize.php
	 * @return mixed data which can be serialized by <b>json_encode</b>,
	 * which is a value of any type other than a resource.
	 */
```

Comment (Raydiation) [inline lib/private/l10n/string.php]: Hm, imho not needed. You also dont add docblocks for __toString and __get.

File: lib/private/l10n/string.php (hunk 1 above)
Comment (Raydiation) [inline lib/private/l10n/string.php]: This is slightly faster :)

File: lib/private/l10n/string.php (hunk 2 above)
Comment (th3fallen) [inline lib/private/l10n/string.php]: I guess you're right on second thought my ide will pickup the docblock of it from the interface anyway.

**Comment by DeepDiver1975** [comment]
another solution would actually be to return string on IL10N::t() and n() ...

**Comment by Raydiation** [comment]
I suppose this was built for lazy evaluating l10n strings.

**Comment by DeepDiver1975** [comment]
> I suppose this was built for lazy evaluating l10n strings.

yes it was - but the benefit is questionable

**Comment by nickvergessen** [comment]
https://github.com/owncloud/core/issues/14663 see point 6
I was about to start on that ticket when it got moved back :(

**Comment by Raydiation** [comment]
Closed in favor of #14663

**Comment by DeepDiver1975** [comment]
well - this change might still be of value for the time being?

@nickvergessen objections? THX

**Comment by nickvergessen** [comment]
No, would be fine by me

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by PVince81** [comment]
Regression on master ?
```
{"reqId":"2e93MMwnlQVFnuvi7a03","remoteAddr":"127.0.0.1","app":"PHP","message":"The use statement with non-compound name 'JsonSerializable' has no effect at \/srv\/www\/htdocs\/owncloud\/lib\/private\/l10n\/string.php#27","level":3,"time":"2015-04-30T11:40:08+00:00","method":"GET","url":"\/owncloud\/index.php\/apps\/files_versions\/download.php?file=%2Fsub%2Fversions.txt&revision=1430393930"}
```

**Comment by PVince81** [comment]
Raised here: https://github.com/owncloud/core/issues/15969

**Comment by nickvergessen** [comment]
Duplicate to https://github.com/owncloud/core/issues/15962

**Pull request merged**